File: 350-IntersectionOfTwoArrII/arrayIntersection2.java
public class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        // use a map to count frequencies.
        Map<Integer, Integer> freq = new HashMap<>();
        List<Integer> rslt = new ArrayList<>();
        
        for (Integer num : nums1) {
            int numFreq = freq.getOrDefault(num, 0);
            freq.put(num, numFreq + 1);
        }
        
        for (Integer num : nums2) {
            int numFreq = freq.getOrDefault(num, 0);
            if (numFreq > 0) {
                rslt.add(num);
                freq.put(num, numFreq - 1);
            }
            
        }
        
        //change rslt to an array
        int[] arr = new int[rslt.size()];
        for (int i = 0; i< arr.length; i++){
            arr[i] = rslt.get(i);
        }

        return arr;
        
        
    }
}

File: 350-IntersectionOfTwoArrII/arrayIntersection2.cpp
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        vector<int> res;
        unordered_map<int,int> mp;
        for(int num:nums1){
            mp[num]++;
        }
        for(int num:nums2){
            if(mp.count(num)&&mp[num]>0){
                mp[num]--;
                res.push_back(num);
            }
        }
        return res;
    }
};